### DISCUSSION TITLE
How to Run the Server from Source

### DISCUSSION THREAD

**Discussion by mtanco** [h2oai/wave#715] [Q&A] [answered]
I am attempting to build and run Wave on an EC2 instance running Ubuntu 20.04 and following the requirements and steps from the Contributing: https://github.com/h2oai/wave/blob/master/.github/CONTRIBUTING.md

When running `make all` the following error has come up:
> string.prototype.trimend not accessible from es-abstract

Below is the full error long and the steps I ran on a clean instance. I see the same error with and without `sudo`. Thank you for any suggestions on what I am missing!

```bash
ubuntu@ip-10-0-0-101:~/wave$ sudo make all
rm -rf build
cd ui && make clean
make[1]: Entering directory '/home/ubuntu/wave/ui'
rm -rf build node_modules
make[1]: Leaving directory '/home/ubuntu/wave/ui'
cd py && make clean
make[1]: Entering directory '/home/ubuntu/wave/py'
rm -rf build dist h2o_wave.egg-info
rm -rf docs/build venv
make[1]: Leaving directory '/home/ubuntu/wave/py'
cd tools/wavegen && make clean
make[1]: Entering directory '/home/ubuntu/wave/tools/wavegen'
rm -rf build node_modules
make[1]: Leaving directory '/home/ubuntu/wave/tools/wavegen'
rm -f waved
cd ui && make setup
make[1]: Entering directory '/home/ubuntu/wave/ui'
npm ci
npm ERR! string.prototype.trimend not accessible from es-abstract

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-03-29T21_56_33_083Z-debug.log
make[1]: *** [Makefile:3: ci] Error 1
make[1]: Leaving directory '/home/ubuntu/wave/ui'
make: *** [Makefile:10: setup] Error 2
```
[npm_logs_2021-03-29T21_56_33_083Z-debug.log ](https://github.com/h2oai/wave/files/6224837/npm_logs_2021-03-29T21_56_33_083Z-debug.log)

Setup Script
```bash

sudo apt update
sudo apt install nodejs
sudo apt install npm

wget -c https://dl.google.com/go/go1.14.2.linux-amd64.tar.gz -O - | sudo tar -xz -C /usr/local
export PATH=$PATH:/usr/local/go/bin

sudo add-apt-repository ppa:deadsnakes/ppa
sudo apt install python3.7
sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.7.10
# add to bashrc: alias python3=python3.7

git clone https://github.com/h2oai/wave.git
cd wave
sudo apt install make
make all
```

**Comment by lo5** [comment, marked as answer]
> sudo apt install nodejs
> sudo apt install npm

Not sure what those lines did. Install the official binaries like this: https://github.com/nodesource/distributions/blob/master/README.md#installation-instructions

Also, there should be no need to downgrade to Python 3.7. Your 20.04 should already have 3.8.

**Comment by mtanco** [comment]
Thanks for the node instructions, I will try that!

The contributing guidelines said 3.7 specifically so good to know 3.8 works 👍🏻

**Reply by lo5** [reply]
3.6+ should be good. Fixed CONTRIBUTING.md.

**Reply by mtanco** [reply]
Thanks!
